[kbss-cvut/termit-ui#521] Allow setting primary language on vocabulary level

lukaskabc opened the pull request:
Resolves kbss-cvut/termit-ui#521

- Fixes `millisToStringConvertsSpecifiedValueToSecondsAndRounds` test to use dynamic decimal separator, allowing it to succeed on systems using non-dot decimal separator (e.g., Czech locale)
- Fixes concurrent modification exception when removing document with multiple files
- Adds `dcterms:language` attribute to Vocabulary defining the primary language of the vocabulary
- When exporting a glossary, the `dcterms:language` will be included
- Validation:
  - The vocabulary and all its terms have to have defined label in the primary language of the vocabulary
  - It is possible to change the language of an existing vocabulary with terms, but no modification to a term can be saved until the missing label is defined
  - The `PrimaryNotBlank` annotation is now only for classes implementing the `HasPrimaryLanguage` interface
- `TermRepositoryService` explicitly inserts vocabulary language in `prepareTermForPersist`, allowing the validation to take place before persisting a new term
- Text analysis is now triggered in the vocabulary primary language, propagating an error to the client via websocket if the language is not supported by the analysis service
- The system will assign the instance language as the primary language to all vocabularies that are missing one on startup

lukaskabc committed:
[Fix] millisToStringConvertsSpecifiedValueToSecondsAndRounds test failing on systems using locale with non-dot decimal separators

Replaces hardcoded dot decimal separator with separator dynamically loaded from locale.

lukaskabc committed:
[kbss-cvut/termit-ui#521] Add primary language attribute to vocabulary model

lukaskabc committed:
[kbss-cvut/termit-ui#521] Update AssetDao to consider vocabulary primary language for labels

lukaskabc committed:
[kbss-cvut/termit-ui#521] Update BaseAssetDao to consider vocabulary primary language for labels

lukaskabc committed:
[kbss-cvut/termit-ui#521] Extract helper pattern methods to helper class SparqlPatterns and update DataDao to consider entity language for labels

lukaskabc committed:
[kbss-cvut/termit-ui#521] Update full text search with per vocabulary language

lukaskabc committed:
[kbss-cvut/termit-ui#521] Create migration service VocabularyPrimaryLanguageGenerator that generates primary language for vocabularies that are missing it

lukaskabc committed:
[kbss-cvut/termit-ui#521] Fix sparql attributes to use DC.Terms.LANGUAGE constant and fix SPARQL queries using literals in COALESCE

lukaskabc committed:
[kbss-cvut/termit-ui#521] Fix full text search missing optional blocks for language

lukaskabc committed:
[kbss-cvut/termit-ui#521] Simplify queries in TermDao regarding vocabulary language selection

lukaskabc committed:
[kbss-cvut/termit-ui#521] Update tests to use vocabulary primary language and mark vocabulary primary language as simple literal string

lukaskabc committed:
[kbss-cvut/termit-ui#521] Add vocabulary language to glossary export

lukaskabc committed:
[kbss-cvut/termit-ui#521] Add vocabulary language loading to SKOSImporter and add language to testing data

lukaskabc committed:
[kbss-cvut/termit-ui#521] Drop SPARQL attributes from terms for vocabulary primary language and replace comparator in TermDao with database query ordering

lukaskabc committed:
[kbss-cvut/termit-ui#521] Rewrite PrimaryNotBlank annotation and its validation

lukaskabc committed:
[kbss-cvut/termit-ui#521] Update services to respect primary language per vocabulary and update their tests to reflect changes. Update constraint violation message from MultilingualStringPrimaryNotBlankValidator.

lukaskabc committed:
[kbss-cvut/termit-ui#521] Use vocabulary language for SKOS importing and document initialization. Fix tests, remove unnecessary stubs.

lukaskabc committed:
[kbss-cvut/termit-ui#521] Update asset daos with vocabulary languages matching labels with vocabulary language or no language

lukaskabc committed:
[kbss-cvut/termit-ui#521] Set vocabulary primary language on each term on prepare term for persist

lukaskabc committed:
[kbss-cvut/termit-ui#521] Set primary language for term before validating during update

lukaskabc committed:
[kbss-cvut/termit-ui#521] Fix failing test for empty term label validation

lukaskabc committed:
[kbss-cvut/termit-ui#521] Update primary not blank validation message and fire vocabulary content modified event when vocabulary itself is modified

lukaskabc committed:
[kbss-cvut/termit-ui#521] Update resource service with file language fallback to vocabulary primary language

lukaskabc committed:
[kbss-cvut/termit-ui#521] Fix tests with updated primary not blank validation message and excel importer test with vocabulary primary language for URI generation

lukaskabc committed:
[kbss-cvut/termit-ui#521] Implement overrides for term and vocabulary primary label and fix wrong method usage for vocabulary primary language

lukaskabc committed:
[kbss-cvut/termit-ui#521] Publish text analysis failure to client websocket

lukaskabc committed:
[kbss-cvut/termit-ui#521] Prepend terms in vocabulary listing that have label in the vocabulary language

lukaskabc committed:
[kbss-cvut/termit-ui#521] Update TextAnalysisServiceTest after adding event publish of TextAnalysisFailedEvent

lukaskabc committed:
[kbss-cvut/termit-ui#521] Update term distribution statistics with vocabulary language for label selection

lukaskabc committed:
[Fix] Prevent concurrent modification exception when deleting a vocabulary with multiple document files by making a copy of the file list

lukaskabc committed:
[kbss-cvut/termit-ui#521] Review changes and minimize diff and optimize queries

lukaskabc committed:
[kbss-cvut/termit-ui#521] Update resource label resolving with vocabulary language for term occurrence info

lukaskabc committed:
[kbss-cvut/termit-ui#521] Add tests for SKOSImporter & DataDao

lukaskabc committed:
[kbss-cvut/termit-ui#521] Add test for concurrent exception while removing document with multiple files

lukaskabc committed:
[kbss-cvut/termit-ui#521] Review changes and refactor for vocabulary getPrimaryLabel usage, improve readability

lukaskabc committed:
[kbss-cvut/termit-ui#521] Fix failing test and fix invalid query for term type distribution missing language attribute in group by statement

lukaskabc committed:
[Fix] Resolve jopa related issue TODO

ledsoft commented on src/main/resources/query/statistics/termDistribution.rq:
Could you rewrite the query to use english variable names? This way it will be consistent with the `termTypeDistribution.rq` file.

ledsoft commented on src/main/java/cz/cvut/kbss/termit/service/init/VocabularyPrimaryLanguageGenerator.java:
```diff
@@ -0,0 +1,55 @@
+    @Transactional
+    public void generateMissingPrimaryLanguage() {
+        LOG.debug("Generating missing primary languages for vocabularies.");
+        em.createNativeQuery("""
```
I would probably put a `?graph` vocabulary in the WHERE clause, as it may happen that the graph IRI does not coincide with the vocabulary IRI.

ledsoft commented on src/main/java/cz/cvut/kbss/termit/service/document/TextAnalysisService.java:
```diff
@@ -144,7 +156,8 @@ private TextAnalysisInput createAnalysisInput(File file) {
         );
         input.setVocabularyRepository(repositoryUrl);
-        input.setLanguage(file.getLanguage() != null ? file.getLanguage() : config.getPersistence().getLanguage());
+        String vocabularyLanguage = vocabularyDao.getReference(file.getDocument().getVocabulary()).getPrimaryLanguage();
```
Consider adding a dedicated DAO method for getting vocabulary language. This may be a common enough behavior that could be useful to optimize not to require loading the entity.

ledsoft commented on src/main/java/cz/cvut/kbss/termit/persistence/dao/DataDao.java:
```diff
@@ -167,28 +173,49 @@ public Optional<String> getLabel(URI id) {
+                                        "BIND (str(?label) as ?strippedLabel )." +
+                                        languageOptionalPattern +
+                   // only bind parameter value to the ?labelLanguage variable if the parameter value is present (COALESCE gives wrong results otherwise)
+                   (languageSpecified ? "BIND (?labelLanguageVal as ?labelLanguage) ." : "") +
```
This line could be added to the `if (!languageSpecified`) to form an if-else and here just concat, right?

ledsoft commented on src/main/java/cz/cvut/kbss/termit/validation/PrimaryNotBlank.java:
```diff
@@ -28,10 +29,12 @@
- * Validation constraint ensuring that a {@link cz.cvut.kbss.jopa.model.MultilingualString} contains a non-empty value
- * in the primary language configured for the application.
+ * Validation constraint ensuring that specified {@link cz.cvut.kbss.jopa.model.MultilingualString MultilingualString}
+ * attributes contains a non-empty value in the primary language of the entity.
```
'attribute contains' or 'attributes contain'

ledsoft commented on src/test/java/cz/cvut/kbss/termit/persistence/dao/DataDaoTest.java:
```diff
@@ -189,6 +193,36 @@ void getLabelReturnsEmptyOptionalForIdentifierWithMultipleLabels() {
+    }
+
+    @Test
+    void getLabelReturnLabelInInstanceLanguageWhenVocabularyIsNotFoundAndNoLanguageIsRequested() {
```
'Returns'

ledsoft commented on src/test/java/cz/cvut/kbss/termit/persistence/dao/skos/SKOSImporterTest.java:
I am missing a test that would verify that the `dc:language` value of glossary is set as the primary language of the imported vocabulary.

ledsoft commented on src/test/java/cz/cvut/kbss/termit/service/validation/ThrottlingValidatorTest.java:
```diff
@@ -80,9 +81,10 @@ void setUp() {
     @Test
     void validateUsesPersistenceLanguageForInternationalizedRules() {
```
This test needs to be renamed to reflect the changed behavior.

ledsoft commented on src/main/java/cz/cvut/kbss/termit/persistence/dao/TermDao.java:
```diff
@@ -150,18 +151,28 @@ private Set<TermInfo> loadInverseRelatedTerms(Term term) {
-                                            "} ORDER BY ?inverse", TermInfo.class)
+                                            "FILTER (?inverse NOT IN (?exclude)) . " +
+                                            insertVocabularyPattern("?inverse") +
+                                            insertLanguagePattern("?inverse") +
```
In this particular case I would vote for inlining the patterns. We know how the vocabulary pattern should look like and the multiple `OPTIONAL` parts from the `SparqlPatterns` just bloat the final query. I am fine with all the other uses of the `SparqlPatterns` but this one seems unnecessary as we know we are working with terms and how the references to vocabulary/language look like.

ledsoft commented on src/main/java/cz/cvut/kbss/termit/persistence/dao/util/SparqlPatterns.java:
```diff
@@ -0,0 +1,72 @@
+
+import java.net.URI;
+
+public class SparqlPatterns {
```
Please add Javadoc to the class explaining its purpose.

ledsoft commented on src/main/java/cz/cvut/kbss/termit/persistence/dao/TermDao.java:
```diff
@@ -538,25 +554,21 @@ public List<TermDto> findAllRoots(Vocabulary vocabulary, Pageable pageSpec, Coll
     }
 
     private String orderSentence(String var) {
```
I think we can move this to the `SparqlPatterns` class.

ledsoft reviewed (commented):
Some inline comments, will be testing.

lukaskabc committed:
[Ref] Rewrite termDistribution query with english variables

lukaskabc committed:
[Ref] Do not restrict primary vocabulary language insertion to graphs matching the vocabulary iri

lukaskabc committed:
[Ref] Introduce Dao method for getting vocabulary primary language instead of using getReference

lukaskabc committed:
[Ref] Reformat query concatenation in DataDao

lukaskabc committed:
[Ref] Fix typos, rename validateUsesPersistenceLanguageForInternationalizedRules test to validateUsesVocabularyPrimaryLanguageForInternationalizedRules

lukaskabc committed:
[Ref] Move orderSentence method to SparqlPatterns and add javadoc

lukaskabc committed:
[Ref] Remove generic SPARQL pattern for vocabulary language from loadInverseTermInfo and use specified matching

lukaskabc committed:
[kbss-cvut/termit-ui#521] Add SKOSImporter test ensuring that a vocabulary is imported with the specified primary language

lukaskabc committed:
[kbss-cvut/termit-ui#521] Resolve failing tests.

Added term in vocabulary relation to persistWithParentEvictsParentsSubTermsCache for child term.

Generate real vocabularies with languages for inverse terms instead of just vocabulary uri in term dao tests

Replaced getReference mocks with the new getPrimaryLanguage

lukaskabc committed:
[kbss-cvut/termit-ui#521] Fix VocabularyDao getPrimaryLanguage implementation and create a test for it

ledsoft reviewed (approved):
LGTM
